ouroboros-consensus/src/ouroboros-consensus/Data/SOP/Match.hs

```diff
@@ -109,6 +116,53 @@ instance HAp (Mismatch f) where
       go (f :* _)  (MR fy gx) = MR fy (apFn f gx)
       go Nil       m          = case m of {}
 
+type instance Same (Mismatch f) = Mismatch f
+
+instance (forall x y. LiftedCoercible p p x y)
+      => HTrans (Mismatch p) (Mismatch p) where
+  htrans ::
+       forall proxy c f g xs ys. AllZipN (Prod (Mismatch p)) c xs ys
+    => proxy c
+    -> (forall x y. c x y => f x -> g y)
+    -> Mismatch p f xs
+    -> Mismatch p g ys
+  htrans p t = \case
+      ML fx gy -> ML (coerce fx) $ htrans p t gy
+      MR fy gx | Dict <- tailDict -> MR (hcoerce fy) $ t gx
+        where
+          tailDict :: Dict (AllZip (LiftedCoercible p p) (Tail xs) (Tail ys))
+          tailDict = replaceAllZipConstraint (Proxy @c)
+      MS m     -> MS $ htrans p t m
+
+  -- NOTE(jdral): this code could be replaced by 'unsafeCoerce' (see 'trans_NP'
+  -- or 'trans_NS' for examples), but this would technically sacrifice type
+  -- safety. For now, this version should be sufficient.
+  hcoerce ::
+       forall f g xs ys. AllZipN (Prod (Mismatch p)) (LiftedCoercible f g) xs ys
+    => Mismatch p f xs
+    -> Mismatch p g ys
+  hcoerce = \case
+      ML fx gy -> ML (coerce fx) $ hcoerce gy
+      MR fy gx | Dict <- tailDict -> MR (hcoerce fy) $ coerce gx
+        where
+          tailDict :: Dict (AllZip (LiftedCoercible p p) (Tail xs) (Tail ys))
+          tailDict = replaceAllZipConstraint (Proxy @(LiftedCoercible f g))
+      MS m     -> MS $ hcoerce m
+
+replaceAllZipConstraint ::
+     forall c c' xs ys.
+     (AllZip c xs ys, forall x y. c' x y)
+  => Proxy c -> Dict (AllZip c' xs ys)
+replaceAllZipConstraint _ = go sList sList
+  where
+    go ::
+         forall as bs. AllZip c as bs
+      => SList as -> SList bs
+      -> Dict (AllZip c' as bs)
+    go SNil  SNil  = Dict
+    go SCons SCons = case go (sList @(Tail as)) (sList @(Tail bs)) of
+        Dict -> Dict
```

Comment:
As discussed in the past on Slack, this could be generalized to
```haskell
impliesAllZip ::
     forall c c' xs ys.
     (AllZip c xs ys, forall x y. c x y => c' x y)
  => Proxy c -> Dict (AllZip c' xs ys)
impliesAllZip _ = go sList sList
  where
    go ::
         forall as bs. AllZip c as bs
      => SList as -> SList bs
      -> Dict (AllZip c' as bs)
    go SNil  SNil  = Dict
    go SCons SCons = case go (sList @(Tail as)) (sList @(Tail bs)) of Dict -> Dict
```
(just to make sure that future users of this function are not confused :smile:)

Reply:
Do you mean the name could be generalised? The code is the same as far as I can see

Reply:
The quantified constraint changed from
```haskell
forall x y. c' x y
```
to
```haskell
forall x y. c x y => c' x y
```
where the latter is less general (easier to satisfy).

Reply:
You're right, I overlooked that!